modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsManagerImpl.java

```diff
@@ -278,4 +739,349 @@ private ObjectStatisticsImpl aggregateLocalStatistics(
+            }
+
+            v.locStatistics.add(msg);
+
+            if (v.remainingNodes.isEmpty()) {
+                aggregateCollected(v);
+
+                return null;
+            }
+
+            return v;
+        });**/
+    }
+
+    /**
+     * Send statistics by request.
+     *
+     * @param nodeId Node to send statistics to.
+     * @param msg Statistics request to process.
+     * @throws IgniteCheckedException In case of errors.
+     */
+    private void supplyStatistics(UUID nodeId, StatsGetRequest msg) throws IgniteCheckedException {
+        List<StatsObjectData> data = new ArrayList<>(msg.keys().size());
+        for (StatsKeyMessage keyMsg : msg.keys()) {
+            StatsKey key = new StatsKey(keyMsg.schema(), keyMsg.obj());
+            ObjectStatisticsImpl objStats = statsRepos.getGlobalStatistics(key);
+
+            if (objStats != null)
+                data.add(StatisticsUtils.toObjectData(keyMsg, StatsType.GLOBAL, objStats));
+        }
+        StatsPropagationMessage res = new StatsPropagationMessage(data);
+        ctx.io().sendToCustomTopic(nodeId, TOPIC, res, GridIoPolicy.QUERY_POOL);
+
+        /**List<StatsObjectData> data = new ArrayList<>();
+        for (StatsKeyMessage key : msg.keys()) {
+            StatsKey statsKey = new StatsKey(key.schema(), key.obj());
+            ObjectStatisticsImpl objStats = statsRepos.getGlobalStatistics(statsKey);
+
+            if (objStats != null)
+                data.add(StatisticsUtils.toMessage(statsKey, StatsType.GLOBAL, objStats));
+        }
+
+        StatsPropagationMessage res = new StatsPropagationMessage(msg.reqId(), data);
+        ctx.io().sendToCustomTopic(nodeId, TOPIC, res, GridIoPolicy.QUERY_POOL);**/
+    }
+
+    private void cancelStatisticsCollection(UUID nodeId, CancelStatsCollectionRequest msg) {
+        currCollections.updateCollection(msg.colId(), stat -> {
+            stat.doneFut().cancel();
+            return null;
+        });
+        /*StatCollectionStatus currState = currCollections.remove(msg.reqId());
+        if (currState != null) {
+            if (log.isDebugEnabled())
+                log.debug(String.format("Cancelling statistics collection by reqId = %s from node %s", msg.reqId(),
+                        nodeId));
+
+            currState.doneFut().cancel();
+        } else
+        if (log.isDebugEnabled())
+            log.debug(String.format("Unable to cancel staitstics collection by req = %s from node %s", msg.reqId(),
+                    nodeId));*/
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onMessage(UUID nodeId, Object msg, byte plc) {
+        try {
+            if (msg instanceof StatsPropagationMessage)
+                receivePartitionsStatistics(nodeId, (StatsPropagationMessage) msg);
+            else if (msg instanceof StatsCollectionResponse)
+                receiveLocalStatistics(nodeId, (StatsCollectionResponse) msg);
+            else if (msg instanceof StatsGetRequest)
```

Comment:
As I understand StatsGetRequest is just return current aggregated statistics and doesn't trigger gathering new stats, right?
This is a different floew and may be we need a separate component for this.

Reply:
Moreover Statistics processor can be started on client node, but it could process only StatGet responses while other messages are not relevant for client node, right?

Reply:
I'm not sure about it, we can run statistics collection from client node and so StatisticsManager will handle most of it's messages even on client. But then we should send aggregated statistics back to some server node.